Refactor create game form

immdevrov opened the pull request:
Hello!
As @kberg told me, a lot of repeated inputs in _CreateGameForm.vue_ needs to be encapsulated.

I've introduced a component called _CreateGameToggle.vue_, which do the trick – hiding input, and give two slots for additional icons. Also, I've extracted tooltip component, I think it will be good for reusing in the future.
 
And, had to add “es2015” lib to ts.config for better autocompletion.

I actually have a lot of other thoughts about how the form can be improved, but  I think this is a proper start

immdevrov commented:
Oh, looks like, npm automatically changed package-lock.json lockfileVersion to 1, pls tell me if it is a critical change – I'll reverte it

kberg commented on src/styles/debug_ui.less:
```diff
@@ -4,7 +4,9 @@
         flex-flow: row wrap !important;
         padding: 20px 10px;
         .expansion-button {
+          & label {
```
What does this do for the create game change?

immdevrov replied:
O, because I moved a class _.expansion-button_ from <label> to <div>, so had to change all the usages across an app.
I thought if I didn't do it, something may crack into other app parts.
Right question here – why this is using the same class as in create game form, if this is not related?

kberg replied:
Because our CSS is a bit sloppy and that UI component is a bit sloppy. Mea culpa. OK, thank you, this makes sense!

kberg commented on .gitignore:
```diff
@@ -16,6 +16,7 @@ pid
 .eslintcache
 tsconfig.tsbuildinfo
 assets/locales
+.idea
```
Moved this up a little.

kberg commented on src/client/components/create/CreateGameForm.vue:
```diff
@@ -1,440 +1,492 @@
+              </template>
+            </CreateGameToggle>
+
+            <CreateGameToggle v-model="promoCardsOption" title="Promos" class="expansion-button">
```
Why is promocards here, and above?

kberg commented on src/client/components/create/CreateGameForm.vue:
```diff
@@ -1,440 +1,492 @@
+
+            <div class="create-game-subsection-label" v-i18n>Fan-made</div>
+
+            <CreateGameToggle v-model="aresExtension" title="Ares" class="expansion-button">
```
So this is a great start, because puts most of the work into the components, but this makes individual checkboxes larger, not smaller. Can this be turned into something that reduces the overall size of a component, like this?

<CreateGameToggle v-model="aresExtension" title="Ares" icon="ares" tooltip="http://....">
</CreateGameToggle>

And then icon can be expanded into the full expansion icon name, and the optional tooltip can be appended.

immdevrov replied:
Yep, I have some thoughts how to reduce it, it is just a first step. I thought It will be better to divide and conquer.

For example. To reduce size of a component, we have to slice it up into few different forms, which will contain some parts of form (divide a big form into a component which control several smaller ones).
Secondly, I think there is a way for using loop render (v-for), which can help to make the form drastically smaller

kberg commented on src/client/components/create/CreateGameToggle.vue:
```diff
@@ -0,0 +1,31 @@
+  },
+  data() {
+    return {
+      inputId: 'i' + (nanoid(10)),
```
Since this is the only purpose for nanoid, I suggest removing it, and relying on the id prop only. For example,

`<label :for="id + '-checkbox'">`

Not only does that remove the change to package-lock.json, but uses the id property which for the moment is unused.

immdevrov replied:
Oh, I just forgot to remove id property, in my opinion adding a new prop just overkill, we have to keep components simpler — we don't need to know this needs _id_, and should let the component decide this.
And nanoid is very VERY lightweight library (look at it like it helper function). [Its weight is about 500 bytes gzipped](https://bundlephobia.com/package/nanoid@3.1.30) and there is no way it'll cause some problems with performance.

kberg replied:
We've adopted a pretty strong "no new components unless necessary" approach, mostly to limit depednencies and rely on active projects as best as possible. This one has no dependencies and _is_actively maintained, but. It's only being used for a form id, which could be replaced with `Math.random().toString()`

![image](https://user-images.githubusercontent.com/413481/148331323-10f1861f-9a04-44b4-b177-bfdd4e3ff735.png)

Don't get me wrong, it's not pretty, but it's essentially the same thing. (Not to mention our existing mechanism for generating unique numbers which you can see in `GameHandler`:

```
  public generateRandomId(prefix: string): string {
    // 281474976710656 possible values.
    return prefix + Math.floor(Math.random() * Math.pow(16, 12)).toString(16);
  }
```

Besides that, though, I'd rather keep the more readable externally-supplied id (or as I was suggesting before, id fragment), please.

immdevrov replied:
If using Math.random is the common way to deal with it in the project – it is ok, I convert to it.

kberg replied:
Sorry I wasn't clear, but the last sentence of my long comment said that actually what I'm asking for is to just supply a readable ID like what was supplied before. I wrote too much about how we didn't need the new module. I shouldn't have written so much.

immdevrov replied:
Why do you need a readable ID? It is one more unnecessary prop here.

kberg reviewed (changes requested):
Great start! I have a couple other observations, which I'll add below.

kberg commented:
This is what the initial page looks like in production:
![image](https://user-images.githubusercontent.com/413481/148157781-ea8ffc6f-efd8-48c0-92c5-8c10a67c4856.png)

and this is what your PR generates:

![image](https://user-images.githubusercontent.com/413481/148157810-d820c155-a8a7-4743-9710-ff5fe34c8e08.png)

Two things I see are that Promos is listed twice, and there are two defaults that aren't set (Corporate Era and Show Timers.)

kberg commented:
When clicking "Venus Next" this is what happens to "The Moon":
![image](https://user-images.githubusercontent.com/413481/148157985-e923eff0-d02b-4608-a907-43d5d2801196.png)

kberg commented:
Adding Turmoil includes "Standard Project Variant"

![image](https://user-images.githubusercontent.com/413481/148158083-08a2456d-c4ca-430b-be55-b5d316223193.png)

kberg commented:
A brief test with Turmoil, Ares and Elysium generated exactly that, which verifies that the model is doing the right things. Great!

![image](https://user-images.githubusercontent.com/413481/148158177-b1c6f395-9485-4291-a196-ca37de752847.png)

I think there's not quite that much to do here, this is a great start, but needs cleanup, etc.

immdevrov commented:
A lot of misplacing happened because of not the best knowledge of how interface should look like, thanks for testing, I'll fix it.

kberg commented:
> 

Yes, thank you! I'm sure we can get this working.

immdevrov commented:
About changes: 
- Replaced nanoid with Math.random()
- Fixed styles (wanna keep them without attribute selector)
- Fixed placement for a little bit (div with width instead of spaces)

immdevrov commented:
About package-lock problem.
screen from [npm docs](https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json)

<img width="752" alt="изображение" src="https://user-images.githubusercontent.com/49252096/148418299-3d8ce8df-0a99-479f-a4fd-269fd334ea08.png">

lockVersion: 2 is designed for npm v7, which is designed with node 15.

As I see in .nvmrc project is using node v14, so it is ok to  use npm v6 and lockVersion: 1 (but we can switch ofc)

kberg commented:
> About changes:
> 
> * Replaced nanoid with Math.random()

I'm sorry if it wasn't clear: please use the ID field and have the caller supply an id. I do not want to use random, either.

> * Fixed styles (wanna keep them without attribute selector)
> * Fixed placement for a little bit (div with width instead of spaces)

immdevrov commented:
I am sorry, I will not do it. I think this is wrong decision and as a professional I do not wont to contribute a code which I find incorrect. I'm open to the conversation, but you have all rights to stand your ground, this is your project.

Furthermore, I found a lot of bad code in a project (bad from my perspective ofc, and I understand this is just my opinion) and now I see why. I do not want to sound disrespectful, actually I admire your work to make my favorite board game available online, and you did a great job (better than steam version), and as a user I appreciate it. But as a software developer, I believe we can not work together.

I am genially sorry, and decide to write this to leave the things clear, instead of just disappearing. So it's ok for you to close merge, or finish it by yourself if you want.

Have a nice day and good luck with the project!

kberg commented:
> I am sorry, I will not do it. I think this is wrong decision and as a professional I do not want to contribute a code which I find incorrect. I'm open to the conversation, but you have all rights to stand your ground, this is your project.

Reading your message, I do not think you are open to a conversation. In fact, you seem pretty closed to it.

> Furthermore, I found a lot of bad code in a project (bad from my perspective ofc, and I understand this is just my opinion) and now I see why. I do not want to sound disrespectful, actually I admire your work to make my favorite board game available online, and you did a great job (better than steam version), and as a user I appreciate it. But as a software developer I believe we can not work together.

Yes, there is a lot of messy code in this project. *A LOT.* I've made plenty of them, but keep in mind:

* The majority of people who started this project (it wasn't me - I joined over a year into its development) were not professional software developers. Many were hobbyists who did not take best practices, latency, reperformance into consideration.
* They chose Vue because they _didn't_ know it, and wanted to learn it.

And yet, with those characteristics, they still managed to create a fully-featured game. (By the time I got on board, all the main expansions were complete!) I'm not trying to say I don't write my share of bad code, but it's just not on the front-end. (I'm still embarrassed by `_AresSetup`.)

Should every PR fix every small thing? No. Should a brand new contributor get their way with what is arguably a minor offense? No.

Here's something you can think about: can we not work together because my standards are too low, or because your expectations are too high? Imagine what we could have accomplished if you ignored that one thing.

bafolts commented:
I was glad to see that @kberg had requested we avoid taking on a new dependency to generate an id. When I originally started the project I had a goal to use NO dependencies other than typescript. If you go back to the early code you will see a very minimal package.json file. Usually every javascript project I work in ends up bogged down by 100+ megabytes of dependencies that have to be downloaded. I think this lack of dependencies is one of the reasons we have ran over 100k games on a hobby plan heroku server with minimal downtime.

I am actually quite proud of the state of the code given it has only been worked on pro bono by strangers. It hasn't been too painful to add all the new expansions as the game expanded. This project is almost 4-5 years old now! For a project of it's age I think it is in great shape. Usually as a project gets to this age it can be difficult to change anything without breaking other parts of the system unexpectedly. The front end is definitely the weaker part of the code but can be more challenging to update as often times one would have to update it all at once. For example if we wanted to convert to react or some other front end framework besides vue-js one would have to either have a hybrid system where some components are written in vue-js and others in react or write every component in react and then cut over. Work like this can be difficult to setup pro bono as we are.

immdevrov commented:
This is your project and it's working and this is the only thing which matters
 Good luck!
